display: stm32_ltdc: rework blanking_on/off

avolmat-st opened the pull request:
The display_blanking_on description states that if available, backlight should also be disabled / enabled on driver blanking on / off. Moreover, upon driver initialization, the display blanking (hence backlight state as well) should be same as if display_blanking_on had been called.

For those reasons, keep the backlight off upon initialization of the driver. Moreover, the LTDC driver does not always have a display_controller attached to it, hence do not return an error when there is no display_controller.

------

I've successfully tested this on STM32MP13xx using the samples/driver/display and on STM32N6 using the samples/driver/video/capture however it would be great if other platforms could be tested as well.

Commit:
display: stm32: ltdc: handle backlight on blanking

The display_blanking_on description states that if available,
backlight should also be disabled / enabled on driver blanking on / off.
Moreover, upon driver initialization, the display blanking (hence
backlight state as well) should be same as if display_blanking_on
had been called.

If nothing can be done by the LTDC driver (no backlight and no
panel blanking) then -ENOSYS is returned.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

avolmat-st commented on drivers/display/display_stm32_ltdc.c:
```diff
@@ -264,11 +264,19 @@ static int stm32_ltdc_display_blanking_off(const struct device *dev)
-	if (display_dev == NULL) {
-		LOG_ERR("There is no panel controller to forward blanking_off call to");
-		return -ENOSYS;
+	if (!display_dev) {
```
Thanks for your comment. I understand your point.

I originally looked at this point because I had issues with some of the sample apps which couldn't work due to the display returning an error on display_blanking_off calls. In turns out that some sample do check the return value (in which case it won't work with display controller not implementing the blanking) and some ignore it.

samples/boards/nordic/nrf_led_matrix/src/main.c:                ret = display_blanking_off(dev);
samples/boards/nordic/nrf_led_matrix/src/main.c:                        printk("display_blanking_off failed: %u/%d\n",
samples/boards/nordic/nrf_led_matrix/src/main.c:                ret = display_blanking_off(dev);
samples/boards/nordic/nrf_led_matrix/src/main.c:                        printk("display_blanking_off failed: %u/%d\n",
samples/drivers/display/src/main.c:     display_blanking_off(display_dev);
samples/drivers/video/capture/src/main.c:       return display_blanking_off(display_dev);
samples/drivers/video/capture_to_lvgl/src/main.c:       display_blanking_off(display_dev);
samples/modules/lvgl/accelerometer_chart/src/main.c:    display_blanking_off(display_dev);
samples/modules/lvgl/demos/src/main.c:  display_blanking_off(display_dev);
samples/modules/lvgl/screen_transparency/src/main.c:    display_blanking_off(display_dev);
samples/subsys/display/cfb/src/main.c:  display_blanking_off(dev);
samples/subsys/display/cfb_custom_font/src/main.c:      if (display_blanking_off(display) != 0) {
samples/subsys/display/lvgl/src/main.c: display_blanking_off(display_dev);
samples/subsys/input/draw_touch_events/src/main.c:      display_blanking_off(display_dev);
samples/subsys/smf/smf_calculator/src/main.c:   display_blanking_off(display_dev);

I was actually in the first place going to try to make clarify the expected behavior of the display_blanking_off / display_blanking_on calls when the underlying driver does not support that call. The application cannot know if it is ok or not to call those display_blanking_off/on and expect no error.
A solution could be to always check for the result and only display an error message if the error is NOT -ENOSYS.

Regarding the LTDC, blanking IS supported, it is properly handling the BL for example (if we have enough information for that) and if there is something else to do via another controller down the display pipeline, then it will be done (and would return an error if that other controller is actually giving us back an error).
But in some cases, there is just nothing else to do. Take the example of a parallel interface RGB panel, there isn't much more thing to do than turning off the backlight and that's the reason why the display-controller device is NULL within the LTDC. That's why in such condition I didn't consider it as an error.have

JarmouniA replied:
> A solution could be to always check for the result and only display an error message if the error is NOT -ENOSYS.

I agree, it should be handled this way on the application side

> Regarding the LTDC, blanking IS supported, it is properly handling the BL for example (if we have enough information for that) and if there is something else to do via another controller down the display pipeline, then it will be done

Still, if `!config->bl_ctrl_gpio.port && !display_dev`, we should return `-ENOSYS`.

avolmat-st replied:
> > A solution could be to always check for the result and only display an error message if the error is NOT -ENOSYS.
> 
> I agree, it should be handled this way on the application side

Alright, so I understand that all application (and thus all samples) should consider that -ENOSYS is not necessarily an error. Do you agree ? If so then I can add some commits into this PR to take care of that.
 
> > Regarding the LTDC, blanking IS supported, it is properly handling the BL for example (if we have enough information for that) and if there is something else to do via another controller down the display pipeline, then it will be done
> 
> Still, if `!config->bl_ctrl_gpio.port && !display_dev`, we should return `-ENOSYS`.

Alright

JarmouniA replied:
> Alright, so I understand that all application (and thus all samples) should consider that -ENOSYS is not necessarily an error. Do you agree ?

Yes
https://github.com/zephyrproject-rtos/zephyr/blob/21f71dfab40ef3c55d82c2aca73f78023a09c76c/include/zephyr/drivers/display.h#L353

Commit:
samples: video: capture: allow -ENOSYS blanking error

The display blanking API mentions that display_blanking_on
and display_blanking_off should return -ENOSYS if it is not
implemented by the driver. It should not be considered as
an error since it might not be needed in order to have the
display work correctly.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
samples: display: cfb_custom_font: allow -ENOSYS blanking error

The display blanking API mentions that display_blanking_on
and display_blanking_off should return -ENOSYS if it is not
implemented by the driver. It should not be considered as
an error since it might not be needed in order to have the
display work correctly.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

JarmouniA commented:
Tested on STM32H750B-DK.

erwango commented:
@kartben, @JarmouniA: What about adding you in "Drivers: display" ?

JarmouniA commented:
> @kartben, @JarmouniA: What about adding you in "Drivers: display" ?

Okey, I will send a PR.
